Nginx : directive proxy_kernel_network_stack

firesidefox opened the pull request:
1. Add a new directive ```proxy_kernel_network_stack``` :
```
Syntax: 	proxy_kernel_network_stack on | off;
Default: 	proxy_kernel_network_stack off;
Context: 	http, stream, mail, server
```
This directive is available only when ```NGX_HAVE_FF_STACK``` is defined.
Determines whether proxy should go throught kernel network stack or fstack.
2. Update F-Stack_Nginx_APP_Guide.md

Commit:
Nginx :  directive proxy_kernel_network_stack
1. Add a new directive proxy_kernel_network_stack :
    Syntax: 	proxy_kernel_network_stack on | off;
    Default: 	proxy_kernel_network_stack off;
   Context: 	http, stream, mail, server
  This directive is available only when NGX_HAVE_FF_STACK is defined.
  Determines whether proxy should go throught kernel network stack or fstack.
2.Update F-Stack_Nginx_APP_Guide.md

firesidefox committed:
Merge pull request #1 from chadwill/develop

Nginx :directive ```proxy_kernel_network_stack```

whl739 commented on app/nginx-1.11.10/src/event/ngx_event_connect.c:
```diff
@@ -38,7 +38,23 @@ ngx_event_connect_peer(ngx_peer_connection_t *pc)
+        s = real_socket(pc->sockaddr->sa_family, type, 0);
+    } else {
+        s = ngx_socket(pc->sockaddr->sa_family, type, 0);
+    }
```
Is there a graceful way? I know that it is a little difficult here.

firesidefox replied:
Maybe we can add a creation flag ```SOCK_CREKERNEL```(create-by-kernel), so that this part can be implemented in the glue layer‘s socket().
```
#define SOCK_CREKERNEL 0x1000
```
Here, or when we want to create socket by kernel, we code like this : 
```
s = ngx_socket(pc->sockaddr->sa_family, type | SOCK_CREKERNEL , 0);
```
Also, ```fstack_territory``` needs to be changed.

What do you think?

whl739 replied:
It's a better way.
What about defining `SOCK_FSTACK`(create a f-stack socket)? Thus we need to modify sockets that are created by f-stack.
But it's up to you.

whl739 commented on doc/F-Stack_Nginx_APP_Guide.md:
````diff
@@ -42,21 +42,37 @@ first one to start |                    |     |                   |    |
-- the `reload` is not graceful, service will still be unavailable during the process of reloading.
+```
+    Syntax: schedule_timeout time;
+    Default: schedule_timeout 30m;
````
Should this be `30` or `30ms`?

firesidefox committed:
Update F-Stack_Nginx_APP_Guide.md

Commit:
Nginx : add a creation flag SOCK_FSTACK(create-fstack-socket) for socket()
1. `#define SOCK_FSTACK 0x1000`
2. when we want to create socket by fstack, we code like this :
`s = ngx_socket(domain, type | SOCK_FSTACK , protocol);`

whl739 commented on app/nginx-1.11.10/src/event/modules/ngx_ff_module.c:
```diff
@@ -208,11 +209,16 @@ ff_mod_init(const char *conf, int proc_id, int proc_type) {
-     if ((AF_INET != domain) || (SOCK_STREAM != type && SOCK_DGRAM != type)) {
+    /* Remove creation flags */
+	type &= ~SOCK_CLOEXEC;
+	type &= ~SOCK_NONBLOCK;
```
wrong indent.

whl739 commented on app/nginx-1.11.10/src/event/modules/ngx_ff_module.c:
```diff
@@ -223,10 +229,15 @@ socket(int domain, int type, int protocol)
 
+	if (unlikely((type & SOCK_FSTACK) == 0)) {
+		return SYSCALL(socket)(domain, type, protocol);
+	}
```
wrong indent.

Commit:
Nginx: revise indent.
